packages/web-runtime/src/composables/upload/useUpload.ts

```diff
@@ -0,0 +1,132 @@
+import { Route } from 'vue-router'
+import { ClientService } from 'web-pkg/src/services'
+import {
+  useCapabilityFilesTusSupportHttpMethodOverride,
+  useCapabilityFilesTusSupportMaxChunkSize,
+  useClientService,
+  useStore
+} from 'web-pkg/src/composables'
+import { computed, Ref, unref, watch } from '@vue/composition-api'
+import { useActiveLocation } from 'files/src/composables'
+import { isLocationPublicActive } from 'files/src/router'
+import { UppyService } from '../../services/uppyService'
+
+// FIXME: properly extend type from AddFileOptions<TMeta extends IndexedObject<any> = Record<string, unknown>>
+export interface UppyResource {
+  id?: string
+  source: string
+  name: string
+  type: string
+  data: Blob
+  meta: {
+    currentFolder: string
+    relativeFolder: string
+    relativeFilePath: string
+    route: Route
+    tusEndpoint: string
+    webDavPath: string
+  }
+}
+
+interface UploadOptions {
+  uppyService: UppyService
+}
+
+interface UploadResult {
+  createDirectoryTree(files: UppyResource[]): void
+}
+
+export function useUpload(options: UploadOptions): UploadResult {
+  const store = useStore()
+  const publicLinkPassword = computed((): string => store.getters['Files/publicLinkPassword'])
+  const isPublicLocation = useActiveLocation(isLocationPublicActive, 'files-public-files')
+  const clientService = useClientService()
+  const getToken = computed((): string => store.getters.getToken)
+
+  const tusHttpMethodOverride = useCapabilityFilesTusSupportHttpMethodOverride()
+  const tusMaxChunkSize = useCapabilityFilesTusSupportMaxChunkSize()
+  const uploadChunkSize = computed((): number => store.getters.configuration.uploadChunkSize)
+
+  const headers = computed((): { [key: string]: string } => {
+    if (unref(isPublicLocation)) {
+      const password = unref(publicLinkPassword)
+
+      if (password) {
+        return { Authorization: 'Basic ' + Buffer.from('public:' + password).toString('base64') }
+      }
+
+      return {}
+    }
+    return {
+      Authorization: 'Bearer ' + unref(getToken)
+    }
+  })
+
+  const uppyOptions = computed(() => {
+    const isTusSupported = unref(tusMaxChunkSize) > 0
+
+    if (isTusSupported) {
+      return {
+        isTusSupported,
+        tusMaxChunkSize: unref(tusMaxChunkSize),
+        uploadChunkSize: unref(uploadChunkSize),
+        tusHttpMethodOverride: unref(tusHttpMethodOverride),
+        headers: unref(headers)
+      }
+    }
+
+    return { isTusSupported, headers: unref(headers) }
+  })
+
+  watch(
+    uppyOptions,
+    () => {
+      if (unref(uppyOptions).isTusSupported) {
+        options.uppyService.useTus(unref(uppyOptions) as any)
+        return
+      }
+      options.uppyService.useXhr(unref(uppyOptions).headers as any)
```

Comment:
I would prefer to pass the whole object here, instead of just the headers prop - for consistency's sake